contracts/strategies/morpho/MetaMorphoStrategy.sol

```diff
@@ -149,11 +203,194 @@ contract MetaMorphoStrategy is ERC4626Strategy {
+      for (uint256 j = 0; j < _rewardDatas.length; j++) {
+        _rewardData = _rewardDatas[j];
+        if (_rewardData.sender == address(0) || (_rewardData.rewardToken == MORPHO && !morphoTransferable)) {
+          continue;
+        }
+        // for each market find the correct market associated the current rewardToken and calculate the apr 
+        apr += _getRewardApr(_mmVault, _aprData, _rewardData.marketId, _quoteRewards(_rewardData));
+      }
+    }
+  }
+
+  /// @notice return the additional rewards apr for a specific market
+  /// @param _mmVault metamorpho vault
+  /// @param _aprData apr data
+  /// @param _targetMarketId target market id
+  /// @param _rewardsInUnderlyings amount of rewards for the market, in underlyings
+  /// @dev return a value multiplied by 1e18 eg for 2% apr -> 2*1e18
+  /// @return apr additional rewards apr
+  function _getRewardApr(
+    IMMVault _mmVault, 
+    AprData memory _aprData,
+    bytes32 _targetMarketId,
+    uint256 _rewardsInUnderlyings
+  ) internal view returns (uint256 apr) {
+    IMorpho.Market memory _market;
+    IMorpho.Position memory _pos;
+    bytes32 _marketId;
+    uint256 _vaultShare;
+    uint256 _assetsSuppliedByVault;
+    uint256 _totalSupplyAssets;
+
+    // find the correct market associated with the current rewardToken and calculate the apr 
+    for (uint256 m = 0; m < _aprData.withdrawLen; m++) {
+      _marketId = _mmVault.withdrawQueue(m);
+      if (_marketId != _targetMarketId) {
+        continue;
+      }
+      // get Morpho Blue market data
+      _market = MORPHO_BLUE.market(_marketId);
+      // get vault position data in the target market
+      _pos = MORPHO_BLUE.position(_marketId, address(_mmVault));
+      // calc how much of `add` will be added to this market
+      if (_aprData.add > 0) {
+        _aprData.add = _calcMarketAdd(_mmVault, _marketId, _aprData.supplyLen, _aprData.add);
+      }
+      // calc how much of `sub` will be added to this market
+      if (_aprData.sub > 0) {
+        _aprData.sub = _calcMarketSub(_mmVault, _marketId, _aprData.withdrawLen, _aprData.sub);
+      }
+      // get underlyings supplied by the vault in the target market
+      // totalSupplyShares : totalSupplyAssets = supplyShares : assetsSuppliedByVault
+      // => assetsSuppliedByVault = supplyShares * totalSupplyAssets / totalSupplyShares
+      _assetsSuppliedByVault = _pos.supplyShares * _market.totalSupplyAssets / _market.totalSupplyShares + _aprData.add - _aprData.sub;
+      // calculate new totalSupplyAssets with liquidity added/removed
+      _totalSupplyAssets = _market.totalSupplyAssets + _aprData.add - _aprData.sub;
+      // calculate vaultShare (% in EXP_SCALE) of the total market and simulate change of liquidity by using add and sub
+      _vaultShare = _assetsSuppliedByVault * EXP_SCALE / _totalSupplyAssets;
+      // calculate vault rewards apr
+      apr = _rewardsInUnderlyings * _vaultShare * 100 / _totalSupplyAssets;
```

Comment:
I would highly recommend you to do the opposite :  Compute the shares that will be minted/burned with the toAssetsDown/toAssetsUp functions of or SharesMathLib, to compute the new vault shares (and new totalShares) after the deposit/withdraw.
Then you can compute : 
`apr = _rewardsInUnderlyings * newVaultShares * 100 / newTotalShares`;
This will be easier and way more accurate (you will limit the shares / asset conversion and not forger the virtual shares in the conversions)

Reply:
This seems to be related with https://github.com/Idle-Labs/idle-tranches/pull/85#discussion_r1477877344 but with `_vaultShare` I'm not computing the vault shares but rather a % of the total market (where 100% is `1e18`)

Reply:
This is needed just to know how much % of rewards will go towards the vault basically so if vault assets are eg 80% of the total assets of the market, the vault should get 80% of the total rewards for that market

Reply:
> This seems to be related with [#85 (comment)](https://github.com/Idle-Labs/idle-tranches/pull/85#discussion_r1477877344) but with `_vaultShare` I'm not computing the vault shares but rather a % of the total market (where 100% is `1e18`)
Yes but `newVaultShares * 100 / newTotalShares` (or `wDivDown(newVaultShares, newTotalShares)` if you want 100% to be equal to `1e18`) gives you the % of the total market owned by the vault (but with a better precision)

Reply:
Initially I was using shares instead of assets here but `add` and `sub` are in assets terms so I kept using assets for consistency for this method. I get your point about the precision but we only need a 'good enough aproximation', even if we have a final apr which is ± 0.1% (ie apr 1.1% instead of 1%) or even more in absolute terms it's fine for our use case. For this reason I would keep the current implementation ideally